File: src/server/gateway.ts
import { encodeNamespace, isPublicObjectKey, namespace } from '../core';
import { GLOBAL_MANIFEST_KEY } from './global-assets';

import type { GlobalAssetManifest } from './global-assets';
import type { RaildropObjectBody, RaildropStorage, RaildropStoredObject } from './storage';

export interface PublicGatewayConfig {
  storage: RaildropStorage;
  prefix?: string;
}

const noStore = (body: string, status: number): Response =>
  new Response(body, {
    status,
    headers: {
      'Access-Control-Allow-Origin': '*',
      'Cache-Control': 'no-store',
      'Content-Type': 'text/plain; charset=utf-8',
      'Cross-Origin-Resource-Policy': 'cross-origin',
      'X-Content-Type-Options': 'nosniff',
    },
  });

const readKey = (pathname: string, prefix: string): string | null => {
  if (prefix && !pathname.startsWith(prefix)) return null;
  const encodedPath = pathname.slice(prefix.length).replace(/^\/+/, '');
  const encodedSegments = encodedPath.split('/');
  if (encodedSegments.some((segment) => !segment)) return null;
  const decodedSegments = encodedSegments.map((segment) => decodeURIComponent(segment));
  const normalizedSegments = namespace(...decodedSegments);
  if (normalizedSegments.some((segment, index) => segment !== decodedSegments[index])) return null;
  return encodeNamespace(normalizedSegments);
};

const unsafeInlineType = (contentType: string | null): boolean =>
  contentType === 'text/html' ||
  contentType === 'application/xhtml+xml' ||
  contentType === 'image/svg+xml' ||
  contentType === 'application/javascript' ||
  contentType === 'text/javascript';

const metadataDownloadName = (metadata: Record<string, string>): string | null => {
  const encoded = metadata['raildrop-name'];
  if (!encoded) return null;
  try {
    return Buffer.from(encoded, 'base64url').toString('utf8');
  } catch {
    return null;
  }
};

const bodyToText = async (body: BodyInit): Promise<string> => new Response(body).text();

const resolveAlias = async (storage: RaildropStorage, alias: string): Promise<string | null> => {
  if (!/^[a-z0-9][a-z0-9-]{0,119}$/.test(alias)) return null;
  const manifestObject = await storage.get(GLOBAL_MANIFEST_KEY);
  if (!manifestObject) return null;
  try {
    const manifest = JSON.parse(
      await bodyToText(manifestObject.body as BodyInit)
    ) as Partial<GlobalAssetManifest>;
    return manifest.version === 1 ? (manifest.assets?.[alias]?.key ?? null) : null;
  } catch {
    return null;
  }
};

export const createPublicGatewayHandler =
  ({ storage, prefix = '' }: PublicGatewayConfig) =>
  async (req: Request): Promise<Response> => {
    if (req.method !== 'GET' && req.method !== 'HEAD') return noStore('Method not allowed', 405);
    const url = new URL(req.url);
    let key: string;
    try {
      const parsed = readKey(url.pathname, prefix);
      if (!parsed) return noStore('Invalid asset path', 400);
      key = parsed;
    } catch {
      return noStore('Invalid asset path', 400);
    }
    const alias = /^global\/([^/]+)$/.exec(key)?.[1];
    if (alias) {
      const resolved = await resolveAlias(storage, alias);
      if (!resolved) return noStore('Not found', 404);
      key = resolved;
    }
    if (!isPublicObjectKey(key)) return noStore('Not found', 404);
    const range = req.headers.get('range') ?? undefined;
    let object: RaildropStoredObject | RaildropObjectBody | null;
    try {
      object =
        req.method === 'HEAD' && !range ? await storage.head(key) : await storage.get(key, range);
    } catch (error) {
      const status = (error as { $metadata?: { httpStatusCode?: number } }).$metadata
        ?.httpStatusCode;
      return status === 416
        ? noStore('Range not satisfiable', 416)
        : noStore('Asset storage unavailable', 503);
    }
    if (!object) return noStore('Not found', 404);
    if (object.metadata['raildrop-access'] && object.metadata['raildrop-access'] !== 'public') {
      return noStore('Not found', 404);
    }
    const expiresAt = object.metadata['raildrop-expires-at'];
    const expiresAtTimestamp = expiresAt ? Date.parse(expiresAt) : Number.NaN;
    if (expiresAt && (!Number.isFinite(expiresAtTimestamp) || expiresAtTimestamp <= Date.now())) {
      return noStore('Not found', 404);
    }
    const isTemporary =
      object.metadata['raildrop-retention'] === 'temporary' || key.startsWith('public/tmp/');
    const temporaryCacheControl =
      isTemporary && Number.isFinite(expiresAtTimestamp)
        ? `public, max-age=${Math.max(0, Math.floor((expiresAtTimestamp - Date.now()) / 1000))}, must-revalidate`
        : 'no-store';
    const headers = new Headers({
      'Accept-Ranges': 'bytes',
      'Access-Control-Allow-Origin': '*',
      'Cache-Control': alias
        ? 'public, max-age=300, stale-while-revalidate=86400'
        : isTemporary
          ? temporaryCacheControl
          : (object.cacheControl ?? 'public, max-age=31536000, immutable'),
      'Content-Length': String(object.size),
      'Content-Type': object.contentType ?? 'application/octet-stream',
      'Cross-Origin-Resource-Policy': 'cross-origin',
      'X-Content-Type-Options': 'nosniff',
    });
    if (object.etag) headers.set('ETag', `"${object.etag}"`);
    if (object.lastModified) headers.set('Last-Modified', object.lastModified.toUTCString());
    if (object.contentDisposition) headers.set('Content-Disposition', object.contentDisposition);
    else if (unsafeInlineType(object.contentType)) {
      const name = metadataDownloadName(object.metadata) ?? 'download';
      headers.set('Content-Disposition', `attachment; filename="${name.replace(/["\\]/g, '_')}"`);
    }
    if ('contentRange' in object && typeof object.contentRange === 'string') {
      headers.set('Content-Range', object.contentRange);
    }
    const ifNoneMatch = req.headers.get('if-none-match') ?? '';
    const etagMatches = object.etag
      ? ifNoneMatch
          .split(',')
          .map((value) => value.trim().replace(/^W\//, '').replace(/^"|"$/g, ''))
          .some((value) => value === '*' || value === object.etag)
      : false;
    const ifModifiedSince = req.headers.get('if-modified-since');
    const modifiedSinceTimestamp = ifModifiedSince ? Date.parse(ifModifiedSince) : Number.NaN;
    const notModifiedSince =
      Boolean(object.lastModified) &&
      Number.isFinite(modifiedSinceTimestamp) &&
      (object.lastModified?.getTime() ?? Number.POSITIVE_INFINITY) <= modifiedSinceTimestamp;
    if (ifNoneMatch ? etagMatches : notModifiedSince) {
      return new Response(null, { status: 304, headers });
    }
    return new Response(
      req.method === 'HEAD' ? null : 'body' in object ? (object.body as BodyInit) : null,
      {
        status: req.headers.has('range') ? 206 : 200,
        headers,
      }
    );
  };

File: sdk/go/gateway.go
package raildrop

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"regexp"
	"strconv"
	"strings"
	"time"
)

const GlobalManifestKey = "public/global/_raildrop/manifest.json"

type PublicGatewayConfig struct {
	Storage Storage
	Prefix  string
}

func noStoreText(w http.ResponseWriter, body string, status int) {
	w.Header().Set("Access-Control-Allow-Origin", "*")
	w.Header().Set("Cache-Control", "no-store")
	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
	w.Header().Set("Cross-Origin-Resource-Policy", "cross-origin")
	w.Header().Set("X-Content-Type-Options", "nosniff")
	w.WriteHeader(status)
	if status >= 200 && status != 204 && status != 304 {
		_, _ = io.WriteString(w, body)
	}
}

func readGatewayKey(pathname string, prefix string) (string, bool) {
	if prefix != "" && !strings.HasPrefix(pathname, prefix) {
		return "", false
	}
	encodedPath := strings.TrimLeft(pathname[len(prefix):], "/")
	if encodedPath == "" {
		return "", false
	}
	encodedSegments := strings.Split(encodedPath, "/")
	for _, segment := range encodedSegments {
		if segment == "" {
			return "", false
		}
	}
	decodedSegments := make([]string, len(encodedSegments))
	for index, segment := range encodedSegments {
		decoded, err := pathUnescape(segment)
		if err != nil {
			return "", false
		}
		decodedSegments[index] = decoded
	}
	normalizedSegments, err := Namespace(decodedSegments...)
	if err != nil {
		return "", false
	}
	for index, segment := range normalizedSegments {
		if segment != decodedSegments[index] {
			return "", false
		}
	}
	encoded, err := EncodeNamespace(normalizedSegments...)
	if err != nil {
		return "", false
	}
	return encoded, true
}

func unsafeInlineType(contentType *string) bool {
	if contentType == nil {
		return false
	}
	switch *contentType {
	case "text/html", "application/xhtml+xml", "image/svg+xml", "application/javascript", "text/javascript":
		return true
	default:
		return false
	}
}

func metadataDownloadName(metadata map[string]string) (string, bool) {
	encoded, ok := metadata["raildrop-name"]
	if !ok || encoded == "" {
		return "", false
	}
	decoded, err := base64.RawURLEncoding.DecodeString(encoded)
	if err != nil {
		return "", false
	}
	return string(decoded), true
}

var globalAliasPattern = regexp.MustCompile(`^global/([^/]+)$`)

func resolveAlias(storage Storage, alias string) (string, error) {
	object, err := storage.Get(context.Background(), GlobalManifestKey, "")
	if err != nil {
		return "", err
	}
	if object == nil {
		return "", nil
	}
	defer object.Close()
	body, err := io.ReadAll(object.Body)
	if err != nil {
		return "", nil
	}
	var manifest struct {
		Version *int `json:"version"`
		Assets  map[string]struct {
			Key string `json:"key"`
		} `json:"assets"`
	}
	if err := json.Unmarshal(body, &manifest); err != nil {
		return "", nil
	}
	if manifest.Version == nil || *manifest.Version != 1 {
		return "", nil
	}
	entry, ok := manifest.Assets[alias]
	if !ok {
		return "", nil
	}
	return entry.Key, nil
}

func NewPublicGateway(config PublicGatewayConfig) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
		if req.Method != http.MethodGet && req.Method != http.MethodHead {
			noStoreText(w, "Method not allowed", http.StatusMethodNotAllowed)
			return
		}
		key, ok := readGatewayKey(req.URL.EscapedPath(), config.Prefix)
		if !ok {
			noStoreText(w, "Invalid asset path", http.StatusBadRequest)
			return
		}
		isAlias := false
		if match := globalAliasPattern.FindStringSubmatch(key); match != nil {
			isAlias = true
			resolved, err := resolveAlias(config.Storage, match[1])
			if err != nil || resolved == "" {
				noStoreText(w, "Not found", http.StatusNotFound)
				return
			}
			key = resolved
		}
		if !IsPublicObjectKey(key) {
			noStoreText(w, "Not found", http.StatusNotFound)
			return
		}
		rangeHeader := req.Header.Get("Range")
		var (
			object *ObjectBody
			err    error
		)
		if req.Method == http.MethodHead && rangeHeader == "" {
			var stored *StoredObject
			stored, err = config.Storage.Head(req.Context(), key)
			if stored != nil {
				object = &ObjectBody{StoredObject: *stored}
			}
		} else {
			object, err = config.Storage.Get(req.Context(), key, rangeHeader)
		}
		if err != nil {
			if errors.Is(err, ErrRangeNotSatisfiable) {
				noStoreText(w, "Range not satisfiable", http.StatusRequestedRangeNotSatisfiable)
			} else {
				noStoreText(w, "Asset storage unavailable", http.StatusServiceUnavailable)
			}
			return
		}
		if object == nil {
			noStoreText(w, "Not found", http.StatusNotFound)
			return
		}
		defer object.Close()
		if access, ok := object.Metadata["raildrop-access"]; ok && access != "public" {
			noStoreText(w, "Not found", http.StatusNotFound)
			return
		}
		expiresAtRaw := object.Metadata["raildrop-expires-at"]
		var expiresAtMs int64
		hasExpiry := expiresAtRaw != ""
		if hasExpiry {
			parsed, parseErr := ParseISOString(expiresAtRaw)
			if parseErr != nil || parsed.UnixMilli() <= time.Now().UnixMilli() {
				noStoreText(w, "Not found", http.StatusNotFound)
				return
			}
			expiresAtMs = parsed.UnixMilli()
		}
		isTemporary := object.Metadata["raildrop-retention"] == "temporary" || strings.HasPrefix(key, "public/tmp/")
		var cacheControl string
		switch {
		case isAlias:
			cacheControl = "public, max-age=300, stale-while-revalidate=86400"
		case isTemporary:
			if hasExpiry {
				maxAge := (expiresAtMs - time.Now().UnixMilli()) / 1000
				if maxAge < 0 {
					maxAge = 0
				}
				cacheControl = "public, max-age=" + strconv.FormatInt(maxAge, 10) + ", must-revalidate"
			} else {
				cacheControl = "no-store"
			}
		case object.CacheControl != nil && *object.CacheControl != "":
			cacheControl = *object.CacheControl
		default:
			cacheControl = "public, max-age=31536000, immutable"
		}
		w.Header().Set("Accept-Ranges", "bytes")
		w.Header().Set("Access-Control-Allow-Origin", "*")
		w.Header().Set("Cache-Control", cacheControl)
		w.Header().Set("Content-Length", fmt.Sprintf("%d", object.Size))
		contentType := "application/octet-stream"
		if object.ContentType != nil {
			contentType = *object.ContentType
		}
		w.Header().Set("Content-Type", contentType)
		w.Header().Set("Cross-Origin-Resource-Policy", "cross-origin")
		w.Header().Set("X-Content-Type-Options", "nosniff")
		if object.ETag != nil {
			w.Header().Set("ETag", "\""+*object.ETag+"\"")
		}
		if object.LastModified != nil {
			w.Header().Set("Last-Modified", object.LastModified.UTC().Format(http.TimeFormat))
		}
		if object.ContentDisposition != nil {
			w.Header().Set("Content-Disposition", *object.ContentDisposition)
		} else if unsafeInlineType(object.ContentType) {
			name, ok := metadataDownloadName(object.Metadata)
			if !ok {
				name = "download"
			}
			sanitized := strings.Map(func(character rune) rune {
				if character == '"' || character == '\\' {
					return '_'
				}
				return character
			}, name)
			w.Header().Set("Content-Disposition", "attachment; filename=\""+sanitized+"\"")
		}
		if object.ContentRange != nil {
			w.Header().Set("Content-Range", *object.ContentRange)
		}
		ifNoneMatch := req.Header.Get("If-None-Match")
		etagMatches := false
		if ifNoneMatch != "" && object.ETag != nil {
			for _, candidate := range strings.Split(ifNoneMatch, ",") {
				value := strings.TrimSpace(candidate)
				value = strings.TrimPrefix(value, "W/")
				value = strings.Trim(value, "\"")
				if value == "*" || value == *object.ETag {
					etagMatches = true
					break
				}
			}
		}
		notModifiedSince := false
		if ifModifiedSince := req.Header.Get("If-Modified-Since"); ifModifiedSince != "" && object.LastModified != nil {
			if parsed, parseErr := http.ParseTime(ifModifiedSince); parseErr == nil {
				notModifiedSince = !object.LastModified.UTC().Truncate(time.Second).After(parsed.UTC())
			}
		}
		if ifNoneMatch != "" {
			if etagMatches {
				w.WriteHeader(http.StatusNotModified)
				return
			}
		} else if notModifiedSince {
			w.WriteHeader(http.StatusNotModified)
			return
		}
		status := http.StatusOK
		if rangeHeader != "" {
			status = http.StatusPartialContent
		}
		if req.Method == http.MethodHead {
			w.WriteHeader(status)
			return
		}
		if object.Body != nil {
			w.WriteHeader(status)
			_, _ = io.Copy(w, object.Body)
			return
		}
		w.WriteHeader(status)
	})
}
